### PR TITLE
Classifier: Run addMiddleware, onAction and onPatch outside of autorun

### PR DESCRIPTION
**Pull request by eatyourgreens**
#2520 revealed a subtle bug in the way that hooks like `onPatch` are being run in the classifier. `onPatch` was called inside an `autorun` listener, which ran every time a new classification was created. A side effect of this was that a new patch listener was created for each subject classified. So the first subject advanced the subject queue by one subject, the second by two subjects, the third by three subjects and so on.

This PR removes `onPatch` from the subjects store and moves it to the root store, where it can observe changes in the `classifications` sub-tree and call actions in the `subjects` sub-tree.

I've also unwrapped `onAction` and `addMiddleware`, which were being run inside `autorun` listeners, and removed some stubs that were needed to pass tests in #2520. In retrospect, I think those stubs were covering up for the bug mentioned above.

With this PR, `onAction`, `onPatch` and `addMiddleWare` are run only once, after the root store is created. The listeners that they create will then run when their respective triggers are triggered.

I've also removed anonymous functions nested inside other functions. Personally, I find anonymous, nested functions hard to read and debug so I've replaced them with named functions in each case.

[Listeners in MobX State Tree](https://mobx-state-tree.js.org/concepts/listeners)

Package:
lib-classifier

# Review Checklist

## General

- [x] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [x] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

### CODE REVIEW

File: packages/lib-classifier/src/store/FeedbackStore.js [hunk 1]
```diff
@@ -70,17 +60,16 @@ const FeedbackStore = types
...
-          if (call.name === 'advance') {
-            onSubjectAdvance(call, next, abort)
-          } else {
-            next(call)
-          }
-        })
-      }, { name: 'FeedbackStore Subject Middleware autorun' })
-      addDisposer(self, subjectMiddleware)
+    function afterAttach () {
+      addMiddleware(getRoot(self).subjects, _addMiddleWare)
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/store/FeedbackStore.js:64]: Why is feedback adding middleware to the subjects store? This is a bit confusing.

File: packages/lib-classifier/src/store/RootStore.js [hunk 2]
```diff
@@ -49,6 +50,30 @@ const RootStore = types
...
+        annotations.forEach(annotation => self.feedback.update(annotation))
+      }
+    }
+
+    function _onPatch(patch) {
+      // TODO: why are we doing this rather than observe classifications.loadingState for changes?
+      const { path, value } = patch
+      if (path === '/classifications/loadingState' && value === 'posting') {
+        self.subjects.advance()
+      }
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/store/RootStore.js:74]: Could this be replaced by an `autorun` function that reacts to changes in `classifications.loadingState`?

File: packages/lib-classifier/src/store/FeedbackStore.js (hunk 1 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/FeedbackStore.js]: The tests mock feedback in isolation, so `getRoot(self).subjects` is undefined when they run.

**Review by mcbouslog — COMMENTED** [review]
I'm in the process of trying to troubleshoot, but noting on https://local.zooniverse.org:3000/projects/nora-dot-test/planet-finders-beta/classify/workflow/3317 the feedback modal doesn't show when expected (can check subject metadata which will note mock transits for feedback). The classification metadata correctly has the feedback (and related success or failure), but no modal showing feedback.

**Comment by eatyourgreens** [comment]
>I'm in the process of trying to troubleshoot, but noting on https://local.zooniverse.org:3000/projects/nora-dot-test/planet-finders-beta/classify/workflow/3317 the feedback modal doesn't show when expected (can check subject metadata which will note mock transits for feedback). The classification metadata correctly has the feedback (and related success or failure), but no modal showing feedback.

`store.subjects` hasn't been set up yet when the feedback store initialises, so `getRoot(self).subjects` is undefined.

**Comment by eatyourgreens** [comment]
@mcbouslog I've fixed feedback by adding `autorun` back in, so that it now observes `getRoot(self)?.subjects` before adding the middleware. I think the difference, with that model, is that it's adding middleware to another model, so we need to wait for `subjects` to be ready.

A better model might move `feedback` from `store.feedback` to `store.subjects.feedback` or `store.subjects.active.feedback`, so that feedback can react to subjects.

I've been testing this on
https://localhost:8080/?project=nora-dot-test/planet-finders-beta&demo=true
using the support for project slugs from #2551.

**Review by mcbouslog — APPROVED** [review]
This LGTM, nice refactor!

**Pull request merged**